### ISSUE TITLE
evars can't always stop typeclass instance resolution, depending only on instance arg order

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5128]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5128
Date: 2016-10-08 03:59:17 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;
To:   

Last updated: 2016-10-08 03:59:17 +0200

## Bugzilla Comment ID: 14232
Date: 2016-10-08 03:59:17 +0200
From: Jonathan Leivent &lt;<jonikelee@gmail.com>&gt;

In another attempt to work out a way to do
 https://coq.inria.fr/bugs/show_bug.cgi?id=5115
this interesting case turned up (in 8.6 version 4da2131):

Class Foo (A:Type) := foo : True.
Class Bar := bar : True.
Class Baz (n:nat) := baz : True.

Instance Foo1 (n:nat)(a:Bar)(b:Baz n) : Foo nat := I.

Instance Foo2 (a:Bar)(n:nat)(b:Baz n) : Foo bool := I.

Typeclasses eauto := debug.
Goal True.
  let o:=open_constr:(_ -> _ -> _ -> Foo nat) in
  unshelve evar (x:o);[typeclasses eauto|]. (*finds Foo1*)
  
  Fail let o:=open_constr:(_ -> _ -> _ -> Foo bool) in
       unshelve evar (y:o);[typeclasses eauto|]. (*fails to find Foo2*)

In other words, the mere reordering of args between Foo1 and Foo2 makes Foo2 impossible to find without triggering search for a Bar, while Foo1 can be found without triggering search for a Bar.
